`NullPointerException` in `ArmeriaGrpcServerInteropTest`

Found this from CI:

```
com.linecorp.armeria.server.grpc.interop.ArmeriaGrpcServerInteropTest > deadlineExceeded FAILED
    java.lang.NullPointerException
```

No stack trace unfortunately.

It seems like the upstream client expects the server to respond with `CANCELLED` rather than `DEADLINE_EXCEEDED`: https://github.com/grpc/grpc-java/blob/39dc883b342f4ffab175195bb0a2e809f8bcd164/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L654-L667

`deadlineExceeded()` does not fail anymore if I change `GrpcService` to close the stream with `CANCELLED` instead of `DEADLINE_EXCEEDED` here: https://github.com/line/armeria/blob/armeria-0.91.0/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java#L182

However, after changing that, I see an intermittent failure at `deaelineExceededServerStreaming()`:

```
java.lang.AssertionError: 
Expected :DEADLINE_EXCEEDED
Actual   :CANCELLED
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at io.grpc.testing.integration.AbstractInteropTest.deadlineExceededServerStreaming(AbstractInteropTest.java:1148)
```

@anuraaga Any suggestions about the right way to fix this?

--- a/grpc/src/main/java/com/linecorp/armeria/client/grpc/ArmeriaClientCall.java
+++ b/grpc/src/main/java/com/linecorp/armeria/client/grpc/ArmeriaClientCall.java
@@ -69,6 +69,7 @@ import io.grpc.DecompressorRegistry;
 import io.grpc.Metadata;
 import io.grpc.MethodDescriptor;
 import io.grpc.Status;
+import io.grpc.Status.Code;
 import io.netty.buffer.ByteBuf;
 import io.netty.util.ByteProcessor;
 import io.netty.util.ByteProcessor.IndexOfProcessor;
@@ -151,6 +152,7 @@ class ArmeriaClientCall<I, O> extends ClientCall<I, O>
                 new ArmeriaMessageDeframer(this, maxInboundMessageSizeBytes, ctx.alloc()),
                 this);
         executor = callOptions.getExecutor();
+
         req.completionFuture().handle((unused1, unused2) -> {
             if (!ctx.log().isAvailable(RequestLogAvailability.REQUEST_CONTENT)) {
                 // Can reach here if the request stream was empty.
@@ -180,6 +182,26 @@ class ArmeriaClientCall<I, O> extends ClientCall<I, O>
         messageFramer.setCompressor(ForwardingCompressor.forGrpc(compressor));
         prepareHeaders(compressor, metadata);
         listener = responseListener;
+
+        if (callOptions.getDeadline() != null) {
+            final long remainingMillis = callOptions.getDeadline().timeRemaining(TimeUnit.MILLISECONDS);
+            if (remainingMillis <= 0) {
+                final Status status = Status.DEADLINE_EXCEEDED
+                        .augmentDescription(
+                                "ClientCall started after deadline exceeded: " +
+                                callOptions.getDeadline());
+                close(status, new Metadata());
+            } else {
+                ctx.setResponseTimeoutMillis(remainingMillis);
+                ctx.setResponseTimeoutHandler(() -> {
+                    final Status status = Status.DEADLINE_EXCEEDED
+                            .augmentDescription(
+                                    "deadline exceeded after " +
+                                    TimeUnit.MILLISECONDS.toNanos(remainingMillis) + "ns.");
+                    close(status, new Metadata());
+                });
+            }
+        }
 
         final HttpResponse res = initContextAndExecuteWithFallback(
                 httpClient, ctx, endpoint,
@@ -367,8 +389,11 @@ class ArmeriaClientCall<I, O> extends ClientCall<I, O>
     }
 
     private void close(Status status, Metadata metadata) {
+        if (status.getCode() == Code.DEADLINE_EXCEEDED && status.getDescription() == null) {
+            status = status.withDescription("ClientCall was cancelled at or after deadline.");
+        }
         ctx.logBuilder().responseContent(GrpcLogUtil.rpcResponse(status, firstResponse), null);
-        req.abort();
+        req.abort(status.asRuntimeException(metadata));
         responseReader.cancel();
 
         try (SafeCloseable ignored = ctx.push()) {

--- a/grpc/src/main/java/com/linecorp/armeria/client/grpc/ArmeriaClientCall.java
+++ b/grpc/src/main/java/com/linecorp/armeria/client/grpc/ArmeriaClientCall.java
@@ -65,6 +65,7 @@ import io.grpc.ClientCall;
 import io.grpc.Codec.Identity;
 import io.grpc.Compressor;
 import io.grpc.CompressorRegistry;
+import io.grpc.Deadline;
 import io.grpc.DecompressorRegistry;
 import io.grpc.Metadata;
 import io.grpc.MethodDescriptor;
@@ -389,8 +390,12 @@ class ArmeriaClientCall<I, O> extends ClientCall<I, O>
     }
 
     private void close(Status status, Metadata metadata) {
-        if (status.getCode() == Code.DEADLINE_EXCEEDED && status.getDescription() == null) {
-            status = status.withDescription("ClientCall was cancelled at or after deadline.");
+        final Deadline deadline = callOptions.getDeadline();
+        if (status.getCode() == Code.CANCELLED && deadline != null && deadline.isExpired()) {
+            status = Status.DEADLINE_EXCEEDED.augmentDescription(
+                    "ClientCall was cancelled at or after deadline.");
+            // Replace trailers to prevent mixing sources of status and trailers.
+            metadata = new Metadata();
         }
         ctx.logBuilder().responseContent(GrpcLogUtil.rpcResponse(status, firstResponse), null);
         if (status.isOk()) {

--- a/grpc/src/main/java/com/linecorp/armeria/client/grpc/ArmeriaClientCall.java
+++ b/grpc/src/main/java/com/linecorp/armeria/client/grpc/ArmeriaClientCall.java
@@ -393,7 +393,11 @@ class ArmeriaClientCall<I, O> extends ClientCall<I, O>
             status = status.withDescription("ClientCall was cancelled at or after deadline.");
         }
         ctx.logBuilder().responseContent(GrpcLogUtil.rpcResponse(status, firstResponse), null);
-        req.abort(status.asRuntimeException(metadata));
+        if (status.isOk()) {
+            req.abort();
+        } else {
+            req.abort(status.asRuntimeException(metadata));
+        }
         responseReader.cancel();
 
         try (SafeCloseable ignored = ctx.push()) {

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java
@@ -522,6 +522,7 @@ class ArmeriaServerCall<I, O> extends ServerCall<I, O>
         return statusToTrailers(ctx, status, metadata, headersSent);
     }
 
+    // Returns ResponseHeaders if headersSent == false or HttpHeaders otherwise.
     static HttpHeaders statusToTrailers(
             ServiceRequestContext ctx, Status status, Metadata metadata, boolean headersSent) {
         final HttpHeadersBuilder trailers = GrpcTrailersUtil.statusToTrailers(

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
@@ -212,7 +212,7 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
         final ArmeriaServerCall<?, ?> call = startCall(
                 methodName, method, ctx, req.headers(), res, serializationFormat);
         if (call != null) {
-            ctx.setRequestTimeoutHandler(() -> call.close(Status.DEADLINE_EXCEEDED, new Metadata()));
+            ctx.setRequestTimeoutHandler(() -> call.close(Status.CANCELLED, new Metadata()));
             req.subscribe(call.messageReader(), ctx.eventLoop(), WITH_POOLED_OBJECTS);
             req.completionFuture().handleAsync(call.messageReader(), ctx.eventLoop());
         }

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
@@ -213,7 +213,6 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
         final ArmeriaServerCall<?, ?> call = startCall(
                 methodName, method, ctx, req.headers(), res, serializationFormat);
         if (call != null) {
-
             ctx.setRequestTimeoutHandler(() -> call.close(Status.DEADLINE_EXCEEDED, new Metadata()));
             req.subscribe(call.messageReader(), ctx.eventLoop(), WITH_POOLED_OBJECTS);
             req.completionFuture().handleAsync(call.messageReader(), ctx.eventLoop());

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
@@ -184,7 +184,7 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
         final ServerMethodDefinition<?, ?> method = registry.lookupMethod(methodName);
         if (method == null) {
             return HttpResponse.of(
-                    ArmeriaServerCall.statusToTrailers(
+                    (ResponseHeaders) ArmeriaServerCall.statusToTrailers(
                             ctx,
                             Status.UNIMPLEMENTED.withDescription("Method not found: " + methodName),
                             new Metadata(),
@@ -199,7 +199,7 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
                     ctx.setRequestTimeout(Duration.ofNanos(timeout));
                 } catch (IllegalArgumentException e) {
                     return HttpResponse.of(
-                            ArmeriaServerCall.statusToTrailers(
+                            (ResponseHeaders) ArmeriaServerCall.statusToTrailers(
                                     ctx, GrpcStatus.fromThrowable(e), new Metadata(), false));
                 }
             }

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
@@ -35,6 +35,7 @@ import java.util.function.Function;
 import javax.annotation.Nullable;
 
 import org.curioswitch.common.protobuf.json.MessageMarshaller;
+import org.curioswitch.common.protobuf.json.MessageMarshaller.Builder;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -110,6 +111,7 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
     private final int maxOutboundMessageSizeBytes;
     private final boolean useBlockingTaskExecutor;
     private final boolean unsafeWrapRequestBuffers;
+    private final boolean useClientTimeoutHeader;
     private final String advertisedEncodingsHeader;
     @Nullable
     private final ProtoReflectionService protoReflectionService;
@@ -123,10 +125,11 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
                 DecompressorRegistry decompressorRegistry,
                 CompressorRegistry compressorRegistry,
                 Set<SerializationFormat> supportedSerializationFormats,
-                Consumer<MessageMarshaller.Builder> jsonMarshallerCustomizer,
+                Consumer<Builder> jsonMarshallerCustomizer,
                 int maxOutboundMessageSizeBytes,
                 boolean useBlockingTaskExecutor,
                 boolean unsafeWrapRequestBuffers,
+                boolean useClientTimeoutHeader,
                 @Nullable ProtoReflectionService protoReflectionService,
                 int maxInboundMessageSizeBytes) {
         this.registry = requireNonNull(registry, "registry");
@@ -134,6 +137,7 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
         this.decompressorRegistry = requireNonNull(decompressorRegistry, "decompressorRegistry");
         this.compressorRegistry = requireNonNull(compressorRegistry, "compressorRegistry");
         this.supportedSerializationFormats = supportedSerializationFormats;
+        this.useClientTimeoutHeader = useClientTimeoutHeader;
         this.protoReflectionService = protoReflectionService;
         jsonMarshaller = jsonMarshaller(registry, supportedSerializationFormats, jsonMarshallerCustomizer);
         this.maxOutboundMessageSizeBytes = maxOutboundMessageSizeBytes;
@@ -188,15 +192,17 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
                             false));
         }
 
-        final String timeoutHeader = req.headers().get(GrpcHeaderNames.GRPC_TIMEOUT);
-        if (timeoutHeader != null) {
-            try {
-                final long timeout = TimeoutHeaderUtil.fromHeaderValue(timeoutHeader);
-                ctx.setRequestTimeout(Duration.ofNanos(timeout));
-            } catch (IllegalArgumentException e) {
-                return HttpResponse.of(
-                        ArmeriaServerCall.statusToTrailers(
-                                ctx, GrpcStatus.fromThrowable(e), new Metadata(), false));
+        if (useClientTimeoutHeader) {
+            final String timeoutHeader = req.headers().get(GrpcHeaderNames.GRPC_TIMEOUT);
+            if (timeoutHeader != null) {
+                try {
+                    final long timeout = TimeoutHeaderUtil.fromHeaderValue(timeoutHeader);
+                    ctx.setRequestTimeout(Duration.ofNanos(timeout));
+                } catch (IllegalArgumentException e) {
+                    return HttpResponse.of(
+                            ArmeriaServerCall.statusToTrailers(
+                                    ctx, GrpcStatus.fromThrowable(e), new Metadata(), false));
+                }
             }
         }
 
@@ -207,6 +213,7 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
         final ArmeriaServerCall<?, ?> call = startCall(
                 methodName, method, ctx, req.headers(), res, serializationFormat);
         if (call != null) {
+
             ctx.setRequestTimeoutHandler(() -> call.close(Status.DEADLINE_EXCEEDED, new Metadata()));
             req.subscribe(call.messageReader(), ctx.eventLoop(), WITH_POOLED_OBJECTS);
             req.completionFuture().handleAsync(call.messageReader(), ctx.eventLoop());

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java
@@ -35,7 +35,6 @@ import java.util.function.Function;
 import javax.annotation.Nullable;
 
 import org.curioswitch.common.protobuf.json.MessageMarshaller;
-import org.curioswitch.common.protobuf.json.MessageMarshaller.Builder;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -125,7 +124,7 @@ public final class GrpcService extends AbstractHttpService implements HttpServic
                 DecompressorRegistry decompressorRegistry,
                 CompressorRegistry compressorRegistry,
                 Set<SerializationFormat> supportedSerializationFormats,
-                Consumer<Builder> jsonMarshallerCustomizer,
+                Consumer<MessageMarshaller.Builder> jsonMarshallerCustomizer,
                 int maxOutboundMessageSizeBytes,
                 boolean useBlockingTaskExecutor,
                 boolean unsafeWrapRequestBuffers,

--- a/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java
+++ b/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java
@@ -21,6 +21,7 @@ import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkState;
 import static java.util.Objects.requireNonNull;
 
+import java.time.Duration;
 import java.util.Set;
 import java.util.concurrent.Executors;
 import java.util.function.Consumer;
@@ -82,6 +83,8 @@ public final class GrpcServiceBuilder {
     private boolean useBlockingTaskExecutor;
 
     private boolean unsafeWrapRequestBuffers;
+
+    private boolean useClientTimeoutHeader = true;
 
     @Nullable
     private ProtoReflectionService protoReflectionService;
@@ -276,6 +279,19 @@ public final class GrpcServiceBuilder {
     }
 
     /**
+     * Sets whether to use a {@code grpc-timeout} header sent by the client to control the timeout for request
+     * processing. If disabled, the request timeout will be the one configured for the Armeria server, e.g.,
+     * using {@link ServerBuilder#requestTimeout(Duration)}.
+     *
+     * <p>It is recommended to disable this when clients are not trusted code, e.g., for grpc-web clients that
+     * can come from arbitrary browsers.
+     */
+    public GrpcServiceBuilder useClientTimeoutHeader(boolean useClientTimeoutHeader) {
+        this.useClientTimeoutHeader = useClientTimeoutHeader;
+        return this;
+    }
+
+    /**
      * Constructs a new {@link GrpcService} that can be bound to
      * {@link ServerBuilder}. It is recommended to bind the service to a server using
      * {@linkplain ServerBuilder#service(HttpServiceWithRoutes, Function[])
@@ -300,6 +316,7 @@ public final class GrpcServiceBuilder {
                 maxOutboundMessageSizeBytes,
                 useBlockingTaskExecutor,
                 unsafeWrapRequestBuffers,
+                useClientTimeoutHeader,
                 protoReflectionService,
                 maxInboundMessageSizeBytes);
         return enableUnframedRequests ? grpcService.decorate(UnframedGrpcService::new) : grpcService;

--- a/grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTest.java
@@ -44,7 +44,6 @@ import javax.annotation.Nullable;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.ClassRule;
-import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.DisableOnDebug;
@@ -168,6 +167,7 @@ public class GrpcClientTest {
                                        .addService(interceptService)
                                        .setMaxInboundMessageSizeBytes(MAX_MESSAGE_SIZE)
                                        .setMaxOutboundMessageSizeBytes(MAX_MESSAGE_SIZE)
+                                       .useClientTimeoutHeader(false)
                                        .build()
                                        .decorate((client, ctx, req) -> {
                                            final HttpResponse res = client.serve(ctx, req);
@@ -1021,29 +1021,48 @@ public class GrpcClientTest {
         fail("cause must be a ResponseTimeoutException or a StatusException: ", cause);
     }
 
-    // NB: It's unclear when anyone would set a negative timeout, and trying to set the negative timeout
-    // into a header correctly raises an exception. The test has been copied over from upstream to make it
-    // easier to understand the compatibility test coverage - not sure why the gRPC test doesn't fail but it
-    // doesn't seem worth investigating too hard on this one.
-    @Ignore
     @Test
     public void deadlineInPast() {
         // Test once with idle channel and once with active channel
         final TestServiceGrpc.TestServiceBlockingStub stub =
-                Clients.newDerivedClient(
-                        blockingStub,
-
-                        ClientOption.RESPONSE_TIMEOUT_MILLIS.newValue(TimeUnit.SECONDS.toMillis(-10)));
-        stub.emptyCall(EMPTY);
-        Throwable t = catchThrowable(() -> stub.emptyCall(EMPTY));
-        assertThat(t).isInstanceOf(StatusRuntimeException.class);
-        assertThat(((StatusRuntimeException) t).getStatus().getCode());
+                blockingStub.withDeadlineAfter(-10, TimeUnit.SECONDS);
+        assertThatThrownBy(() -> stub.emptyCall(EMPTY))
+                .isInstanceOfSatisfying(StatusRuntimeException.class, t -> {
+                    assertThat(t.getStatus().getCode()).isEqualTo(Code.DEADLINE_EXCEEDED);
+                    assertThat(t.getStatus().getDescription())
+                            .startsWith("ClientCall started after deadline exceeded");
+                });
 
         // warm up the channel
         blockingStub.emptyCall(Empty.getDefaultInstance());
-        t = catchThrowable(() -> stub.emptyCall(EMPTY));
-        assertThat(t).isInstanceOf(StatusRuntimeException.class);
-        assertThat(((StatusRuntimeException) t).getStatus().getCode());
+        assertThatThrownBy(() -> stub.emptyCall(EMPTY))
+                .isInstanceOfSatisfying(StatusRuntimeException.class, t -> {
+                    assertThat(t.getStatus().getCode()).isEqualTo(Code.DEADLINE_EXCEEDED);
+                    assertThat(t.getStatus().getDescription())
+                            .startsWith("ClientCall started after deadline exceeded");
+                });
+    }
+
+    @Test
+    public void deadlineInFuture() throws Exception {
+        final TestServiceGrpc.TestServiceStub stub =
+                asyncStub.withDeadlineAfter(100, TimeUnit.MILLISECONDS);
+        final StreamRecorder<StreamingOutputCallResponse> responseObserver = StreamRecorder.create();
+        stub.streamingOutputCall(
+                StreamingOutputCallRequest
+                        .newBuilder()
+                        .addResponseParameters(
+                                ResponseParameters.newBuilder()
+                                                  .setIntervalUs((int) TimeUnit.SECONDS.toMicros(1))
+                                                  .build())
+                        .build(),
+                responseObserver);
+        responseObserver.awaitCompletion(operationTimeoutMillis(), TimeUnit.MILLISECONDS);
+        assertThat(responseObserver.getError()).isInstanceOfSatisfying(
+                StatusRuntimeException.class, t -> {
+                    assertThat(t.getStatus().getCode()).isEqualTo(Code.DEADLINE_EXCEEDED);
+                    assertThat(t.getStatus().getDescription()).contains("deadline exceeded after");
+                });
     }
 
     @Test

--- a/grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceServerTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceServerTest.java
@@ -747,11 +747,8 @@ class GrpcServiceServerTest {
                     .withDeadlineAfter(10, TimeUnit.SECONDS);
             assertThatThrownBy(() -> client.timesOut(
                     SimpleRequest.getDefaultInstance())).isInstanceOfSatisfying(
-                    StatusRuntimeException.class, t -> {
-                        assertThat(t.getStatus().getCode()).isEqualTo(Code.DEADLINE_EXCEEDED);
-                        assertThat(t.getStatus().getDescription())
-                                .isEqualTo("ClientCall was cancelled at or after deadline.");
-                    });
+                    StatusRuntimeException.class, t ->
+                            assertThat(t.getStatus().getCode()).isEqualTo(Code.CANCELLED));
         });
     }
 

--- a/grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceServerTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceServerTest.java
@@ -30,6 +30,7 @@ import java.util.Arrays;
 import java.util.Base64;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.LinkedTransferQueue;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Stream;
 
@@ -343,6 +344,11 @@ class GrpcServiceServerTest {
             responseObserver.onNext(SimpleResponse.getDefaultInstance());
             responseObserver.onCompleted();
         }
+
+        @Override
+        public void timesOut(SimpleRequest request, StreamObserver<SimpleResponse> responseObserver) {
+            ServiceRequestContext.current().setRequestTimeoutMillis(100);
+        }
     }
 
     private static final ServerInterceptor REPLACE_EXCEPTION = new ServerInterceptor() {
@@ -401,7 +407,7 @@ class GrpcServiceServerTest {
                                 return delegate.serve(ctx, req);
                             }));
 
-            // For simplicity, mount onto a subpath with custom options
+            // For simplicity, mount onto subpaths with custom options
             sb.serviceUnder(
                     "/json-preserving/",
                     GrpcService.builder()
@@ -409,6 +415,12 @@ class GrpcServiceServerTest {
                                .supportedSerializationFormats(GrpcSerializationFormats.values())
                                .jsonMarshallerCustomizer(
                                        marshaller -> marshaller.preservingProtoFieldNames(true))
+                               .build());
+            sb.serviceUnder(
+                    "/no-client-timeout/",
+                    GrpcService.builder()
+                               .addService(new UnitTestServiceImpl())
+                               .useClientTimeoutHeader(false)
                                .build());
 
             sb.service(
@@ -724,6 +736,23 @@ class GrpcServiceServerTest {
                             .setPayload(Payload.newBuilder()
                                                .setBody(ByteString.copyFrom(LARGE_PAYLOAD.toByteArray())))
                             .build();
+    }
+
+    @Test
+    void ignoreClientTimeout() {
+        withTimeout(() -> {
+            final UnitTestServiceBlockingStub client =
+                    new ClientBuilder(server.httpUri(GrpcSerializationFormats.PROTO, "/no-client-timeout/"))
+                    .build(UnitTestServiceBlockingStub.class)
+                    .withDeadlineAfter(10, TimeUnit.SECONDS);
+            assertThatThrownBy(() -> client.timesOut(
+                    SimpleRequest.getDefaultInstance())).isInstanceOfSatisfying(
+                    StatusRuntimeException.class, t -> {
+                        assertThat(t.getStatus().getCode()).isEqualTo(Code.DEADLINE_EXCEEDED);
+                        assertThat(t.getStatus().getDescription())
+                                .isEqualTo("ClientCall was cancelled at or after deadline.");
+                    });
+        });
     }
 
     @Test

--- a/grpc/src/test/proto/com/linecorp/armeria/grpc/testing/test.proto
+++ b/grpc/src/test/proto/com/linecorp/armeria/grpc/testing/test.proto
@@ -143,6 +143,9 @@ service UnitTestService {
 
     // A call where we check an interceptor correctly set a context value.
     rpc GrpcContext(SimpleRequest) returns (SimpleResponse);
+
+    // A call which will not return the response, triggering server timeout.
+    rpc TimesOut(SimpleRequest) returns (SimpleResponse);
 }
 
 // A simple service NOT implemented at servers so clients can test for

--- a/it/server/src/test/java/com/linecorp/armeria/server/grpc/interop/ArmeriaGrpcServerInteropTest.java
+++ b/it/server/src/test/java/com/linecorp/armeria/server/grpc/interop/ArmeriaGrpcServerInteropTest.java
@@ -16,8 +16,6 @@
 
 package com.linecorp.armeria.server.grpc.interop;
 
-import static org.junit.Assume.assumeFalse;
-
 import java.io.IOException;
 import java.net.InetSocketAddress;
 import java.util.concurrent.Executors;
@@ -120,14 +118,6 @@ public class ArmeriaGrpcServerInteropTest extends AbstractInteropTest {
     protected boolean metricsExpected() {
         // Armeria handles metrics using micrometer and does not support opencensus.
         return false;
-    }
-
-    @Override
-    public void deadlineExceeded() throws Exception {
-        // FIXME(trustin): Re-enable this test on Windows once we fix #2008
-        //                 https://github.com/line/armeria/issues/2008
-        assumeFalse(System.getProperty("os.name", "").startsWith("Win"));
-        super.deadlineExceeded();
     }
 
     // This base implementation is to check that the client sends the timeout as a request header, not that the

